File: 5-longest-palindromic-substring/5-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
const longestPalindrome = function(s) {
    let result = '',
        maxLen = 0;
    
    for (let i = 0; i < s.length; i++) {
        
        // odd length palindrome
        let l = i,
            r = i;
        while (l >= 0 && r < s.length && s[l] === s[r]) {
            const currLen = r - l + 1;
            if (maxLen < currLen) {
                maxLen = currLen;
                result = s.slice(l, r + 1);
            }
            l--;
            r++;
        }
        
        // even length palindrome
        l = i;
        r = i + 1;
        while (l >= 0 && r < s.length && s[l] === s[r]) {
             const currLen = r - l + 1;
            if (maxLen < currLen) {
                maxLen = currLen;
                result = s.slice(l, r + 1);
            }
            l--;
            r++;
        }
    }
    return result;
};

// brute force O(n^3)
// above solution O(n^2)

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.rb
# @param {String} s
# @return {String}
def longest_palindrome(s)
    res = ''
    resLen = 0
    (0..s.length-1).each do |i|
        # odd length
        l, r = i, i
        while l >= 0 && r < s.length && s[l] == s[r]
           if (r - l + 1) > resLen
               res = s[l..r]
               resLen = r - l + 1
           end
           l -= 1
           r += 1
        end
        # even length
        l, r = i, i+1
        while l >= 0 && r < s.length && s[l] == s[r]
           if (r - l + 1) > resLen
               res = s[l..r]
               resLen = r - l + 1
           end
           l -= 1
           r += 1
        end
    end
    res
end